### PR TITLE
Add configuration for overriding the RefMode

### PR DESCRIPTION
**Pull request by sdroege**
This is mostly useful for manual types where the actually implemented
RefMode is different from what GIR detects automatically.

In my case, I have GstCaps which is a refcounted type. It is basically a subclass of GstMiniObject (not a GObject), and gst_mini_object_ref/unref has to be used on it. There is no gst_caps_ref/unref, so GIR does not know about this and thinks that it's a RecordType::Direct and then uses to_glib_none_mut() when passing out pointers. However this makes no sense for the way how it's implemented.

With the configuration here, I can override the RefMode to ByRef, and everything works fine.

### CODE REVIEW

File: src/analysis/child_properties.rs [hunk 1]
```diff
@@ -106,8 +106,8 @@ fn analyze_property(
         if conversion != properties::PropertyConversion::Direct {
             imports.add("std::mem::transmute", None);
         }
-        let get_out_ref_mode = RefMode::of(&env.library, typ, library::ParameterDirection::Return);
-        let mut set_in_ref_mode = RefMode::of(&env.library, typ, library::ParameterDirection::In);
+        let get_out_ref_mode = RefMode::of(&env, typ, library::ParameterDirection::Return);
+        let mut set_in_ref_mode = RefMode::of(&env, typ, library::ParameterDirection::In);
```
Comment (EPashkin) [inline src/analysis/child_properties.rs]: &env -> env.
On all uses `RefMode::of` too.

File: src/analysis/parameter.rs [hunk 2]
```diff
@@ -67,7 +67,7 @@ pub fn analyze(
         .matched_parameters(&name)
         .iter()
         .any(|p| p.constant);
-    let ref_mode = RefMode::without_unneeded_mut(&env.library, par, immutable);
+    let ref_mode = RefMode::without_unneeded_mut(&env, par, immutable);
```
Comment (EPashkin) [inline src/analysis/parameter.rs]: just `env` without ampersand too.

File: src/config/gobjects.rs [hunk 3]
```diff
@@ -143,8 +146,25 @@ fn parse_object(toml_object: &Value, concurrency: library::Concurrency) -> GObje
...
+        .lookup("ref_mode")
+        .and_then(|v| v.as_str())
+        .and_then(|v| match v {
+            "none" => Some(ref_mode::RefMode::None),
+            "ref" => Some(ref_mode::RefMode::ByRef),
+            "ref-mut" => Some(ref_mode::RefMode::ByRefMut),
+            "ref-immut" => Some(ref_mode::RefMode::ByRefImmut),
+            "ref-fake" => Some(ref_mode::RefMode::ByRefFake),
+            _ => None,
+        });
```
Comment (EPashkin) [inline src/config/gobjects.rs]: Please move this match out this function. Just private function in this module IMHO will do as `analysis/ref_mode.rs` use this module.

**Comment by EPashkin** [comment]
Also IMHO mostly `to_glib_none_mut` go from non-const in ffi definition and can be fixed with "const=true" on this parameter. See https://github.com/gtk-rs/gtk/blob/master/Gir.toml#L312-L314.

**Comment by EPashkin** [comment]
Maybe this PR will helps removing need of adding many "const=true".

**Comment by EPashkin** [comment]
And please update readme.md too (maybe will conflict with #406)

**Comment by sdroege** [comment]
All done, thanks for the review. I'm not sure this is the right approach, but all other alternatives for doing what I need here seem equally bad :)

**Comment by EPashkin** [comment]
Thanks, I merge this after CI passed.

**Pull request merged**